Add support for custom Envoy JSON fields

mike1808 opened the pull request:
Add support for custom Envoy JSON fields via `=` in the fields name. The specify custom field name and its Envoy format string use `=` in the field name. The syntax is the following: `field_name=Envoy format string`, where Envoy format string can have any command except `DYNAMIC_METADATA` and `FILTER_STATE` commands. 

Example of custom field:

```yaml
json-fields:
  - @timestamp
  - customer_id=%REQ(X-CUSTOMER-ID)%
```

Fixes #3032, Fixes #1507

cc @KauzClay

jpeach commented on pkg/config/parameters_test.go:
```diff
@@ -126,6 +126,24 @@ func TestValidateAccessLogType(t *testing.T) {
 }
 
+func TestValidateAccessLogFields(t *testing.T) {
+	assert.Error(t, AccessLogFields([]string{"dog", "cat"}).Validate())
```
Should we accept something like`[]string{"dog=pug", "cat=black"}`?

mike1808 replied:
@jpeach in theory Envoy supports that and it will put the static value `dog: field` in access log.  I think, if we're supporting other string operators (static strings) we should also accept this.

jpeach commented on pkg/config/parameters.go:
```diff
@@ -89,6 +90,88 @@ func (a AccessLogType) Validate() error {
 const EnvoyAccessLog AccessLogType = "envoy"
 const JSONAccessLog AccessLogType = "json"
 
+type AccessLogFields []string
```
Why don't you make this a `map[string]string` so that it can be fully denormalized in the config? i.e, after validation, break the config down into key/value pairs so that the caller in `internal/envoy/v2/accesslog.go` doesn't need to do it.

mike1808 replied:
We were thinking about that, but because this type is used for a [YAML config parser](https://github.com/projectcontour/contour/pull/3059/files/c007f5c3b7e6ad3edb1c617a57f945ded2e9490f#diff-a324876cd0bc5df1659faac7631005ed9d83b9445c6720142db3c8d2d3e83e56R374) we cannot change it as it will break the current API. One option is to save it to a different field and then copy to this fields as a `map[string]string`

jpeach replied:
> We were thinking about that, but because this type is used for a [YAML config parser](https://github.com/projectcontour/contour/pull/3059/files/c007f5c3b7e6ad3edb1c617a57f945ded2e9490f#diff-a324876cd0bc5df1659faac7631005ed9d83b9445c6720142db3c8d2d3e83e56R374) we cannot change it as it will break the current API.

Oh good point, I had blanked that out. WDYT about adding a method to to the conversion:
```Go
func (a *AccessLogFields) AsFieldMap() map[string]string {
...
}
```

mike1808 replied:
Done

jpeach reviewed (changes requested):
Hi @mike1808,

Overally, I think this looks pretty good.

1. We should document this in `./site/_site/docs/main/configuration.md`. It's
OK to do this in a separate PR if you prefer.

2. I think that it would be clearer to split the validation regex to handle the
simple case (e.g. `%HOSTNAME%`) and complex case (e.g. `%REQ(foo?bar):baz%`)
separately.  This would make the regex much more obvious and the code easier to
follow. The `REQ/RESP` regex only needs to validate the capture from the simple
case (i.e. `REQ(foo?bar):baz`)

3. Once you do (2), there's an opportunity to have a single definition of the
simple expressions. If you define `EnvoyOperators` to just be the simple
operators, then you can derive the JSON field name from that, and add those
mappings to `JSONFields` in an `init` function. The `EnvoyOperators` map should
go into `pkg/config/accesslog.go`.

mike1808 commented on pkg/config/parameters_test.go:
```diff
@@ -126,6 +126,24 @@ func TestValidateAccessLogType(t *testing.T) {
+	assert.NoError(t, AccessLogFields([]string{"@timestamp", "duration=My duration=%DURATION%.0"}).Validate())
+	assert.NoError(t, AccessLogFields([]string{"@timestamp", "duration=%START_TIME(%s.%6f)%"}).Validate())
+	assert.NoError(t, AccessLogFields([]string{"@timestamp", "content-id=%REQ(X-CONTENT-ID)%"}).Validate())
+	assert.NoError(t, AccessLogFields([]string{"@timestamp", "duration=my durations are %DURATION%.0 and method is %REQ(:METHOD)%"}).Validate())
```
Yes, it does.

mike1808 commented:
@jpeach 

1. We'll add the docs in the separate PR.
2. I don't think that's possible because Envoy can have multiple operators in one field. E.g. it can have a simple operator like `%HOSTNAME%` and parameterized one like `%REQ()%`. like `req from %HOSTNAME% with %REQ(X-CONTNENT-ID)%`
3. Because the option (2) is not possible we cannot do that. But we can move `EnvoyOperators` to `pkg/config/accesslog.go`

mike1808 commented:
@jpeach we updated the PR and try to addressed all review comments. One thing we're not sure is our approach for keeping parsed access log fields in `Parameters` structure.

cc @xanderstrike

mike1808 commented on pkg/config/parameters_test.go:
```diff
@@ -126,6 +126,24 @@ func TestValidateAccessLogType(t *testing.T) {
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "dog"}).Validate())
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "content-id=%REQ=dog%"}).Validate())
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "content-id=%dog(%"}).Validate())
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "content-id=%REQ()%"}).Validate())
```
Done

mike1808 commented on pkg/config/parameters_test.go:
```diff
@@ -126,6 +126,24 @@ func TestValidateAccessLogType(t *testing.T) {
+func TestValidateAccessLogFields(t *testing.T) {
+	assert.Error(t, AccessLogFields([]string{"dog", "cat"}).Validate())
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "dog"}).Validate())
+	assert.Error(t, AccessLogFields([]string{"@timestamp", "content-id=%REQ=dog%"}).Validate())
```
Fixed

mike1808 commented:
@jpeach we renamed `JSONFields` to `jsonFields` and encapsulated all the translation logic in `config` package.

youngnick reviewed (approved):
LGTM, nice work. I'm looking forward to seeing the docs update as well, I think that's key for making this feature more usable. We should wait for @jpeach to approve as well though.

jpeach commented on pkg/config/parameters.go:
```diff
@@ -89,6 +90,82 @@ func (a AccessLogType) Validate() error {
+				operator = fmt.Sprintf("%%%s%%", strings.ToUpper(val))
+			}
+
+			fieldMap[val] = operator
```
I think we are missing a case here. There's 3 things that can happen
1. The field name exists in `jsonFields`
2. The field name is a case-folding match to a `envoySimpleOperators` entry
3. Neither of the above.

In case (3), we would basically do `fieldMap[val] = ""`. I think that this is OK, because we subsequently validate the mapping. Let's comment this clearly for future readers!

XanderStrike replied:
This is good feedback thank you. I've reorganized the cases and took another look at the variable names to hopefully clarify what's going on, and of course added some comments to make it extra clear.

jpeach commented on pkg/config/parameters.go:
```diff
@@ -89,6 +90,82 @@ func (a AccessLogType) Validate() error {
+			if op == "START_TIME" && f[4] != "" {
+				return fmt.Errorf("invalid JSON field: %s, invalid Envoy format: %s, arguments required for operator: %s", val, f, op)
+			}
+			// START_TIME cannot not have truncation length
```
Nit: can you please make comments are full sentences, including a period (here and above).

jpeach reviewed (approved):
Thanks @mike1808, this looks great. Could you just take care of the couple of nits, then we can merge.

XanderStrike commented on pkg/config/parameters.go:
```diff
@@ -89,6 +90,82 @@ func (a AccessLogType) Validate() error {
+	for _, val := range a {
+		parts := strings.SplitN(val, "=", 2)
+		if len(parts) == 1 {
+			operator := ""
```
Good catch, addressed

XanderStrike commented:
Latest comments addressed and fixed up into the commit, rebased against `projectcontour:main`. Ready for review again.
